feat(mongodb-compass): Remove loading screen from the app part 1 COMPASS-5113

gribnoysup opened the pull request:
## Description

This PR removes the loading screen from Compass application, showing the main connection window right away instead as soon as the window is ready with the skeleton placeholder for the initial page load. This PR also includes a few additional webpack optimizations for the Compass build:

- use `splitChunks` option so that we get more than one code bundle for the renderer process, this helps parallelize loading, but more importantly parsing of the application code in the renderer
- extract application css to files so that renderer can process it separately from javascript

As a kinda unavoidable drive-by I refactored Compass main code to typescript, I haven't changed the logic much from the previous implementation, just cleaned up obvious issues highlighted by TypeScript and solidified all the logic around singleton instances for every "part" of the main process: menu handling, window management, logging, and auto updating. Also decoupled some event handling by allowing these new singletons to attach events to the `CompassApplication` instance (before this was kinda done by using `electron.app` as a global event bus or directly calling methods from those singletons which is blurs usage patterns) with strictly defined event types.

I am explicitly marking this as part one as I want to use this ticket as an opportunity to clean up the renderer code also, there are some opportunities to speed up start times there also and it should be good for code health of compass in general. But this code change is big enough already and doesn't need anything else (except for the code review / feedback of course) and I want to open it for review before I'm going on a vacation so y'all have enough time to look at it.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

Assumption is that Compass startup times after all the webpack build work are small enough that we don't want to have an explicit loading screen anymore, instead showing the connection screen right away (with a skeleton placeholder for the unavoidable time for application to get completely ready)

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Open Questions

- [ ] Need to check how the app feels now with Claudia to see if we need to tweak something

## Dependents
<!--- If applicable, link PRs/commits that this PR is dependent on or is a dependency of. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

gribnoysup commented on packages/compass/src/main/logging.ts:
```diff
@@ -0,0 +1,227 @@
+
+  static async init(app: CompassApplication): Promise<void> {
+    if (this.initPromise) {
+      await this.initPromise;
```
```suggestion
      await this.initPromise;
      return;
```

lerouxb replied:
I just came here to say the same thing - surely you mean to return here so you don't init twice?

gribnoysup commented on packages/compass/src/main/window-manager.ts:
```diff
@@ -0,0 +1,313 @@
+
+  window.once('closed', onWindowClosed);
+
+  CompassMenu.load(window);
```
Move this to CompassApplication eventbus

gribnoysup commented on packages/compass/tsconfig-lint.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "./tsconfig.json",
+  "include": ["**/*"],
+  "exclude": ["node_modules", "lib"]
```
```suggestion
  "exclude": ["node_modules", "build", "dist"]
```

lerouxb commented on packages/compass/src/main/handle-uncaught-exception.ts:
```diff
@@ -0,0 +1,54 @@
+    });
+
+    /**
+     * TODO (@imlucas) Copy diagnostics to clipboard?
```
Maybe we should use this opportunity to remove TODOs from years past that we're not going to do unless a ticket to do it comes up.

lerouxb commented on packages/compass/src/main/window-manager.ts:
```diff
@@ -0,0 +1,313 @@
+let appLaunched = false;
+
+async function showWindowWhenReady(bw: BrowserWindow) {
+  await once(bw, 'ready-to-show');
```
Is it possible that ready-to-show could have already been emitted before this function is called?

lerouxb replied:
Nevermind I see this is indeed called before window.loadURL which I assume is where that will happen.

lerouxb commented on packages/hadron-auto-update-manager/package.json:
```diff
@@ -20,6 +20,7 @@
     "webpack": "./index.js",
     "require": "./index.js"
   },
+  "types": "index.d.ts",
```
With these index.d.ts files, how much more work would it have been to just make the files typescript rather than provide types for .js files? Or will that break something else that needs it to be .js?

lerouxb commented on packages/compass/src/main/application.ts:
```diff
@@ -0,0 +1,136 @@
+  setupUserDirectory(): void {
+    if (process.env.NODE_ENV === 'development') {
+      const appName = app.getName();
+      // When NODE_ENV is dev, we are probably be running application unpackaged
```
typo. "are probably be running"

gribnoysup commented:
> Also - as we've chatted about a bit, this main part of Compass doesn't have much in the way of unit tests. Is that something we want to add? Will that be in part 2?

@Anemy initially planned to have some tests added in this PR, but will have them now split in a separate task to be able to proceed with merging this change. Added tickets COMPASS-5188 and COMPASS-5189
